pypesto/objective/history.py

```diff
@@ -713,41 +694,59 @@ def _save_trace(self, finalize: bool = False):
                 )
             trace_copy.to_csv(self.file)
 
-    def get_x(self, ix) -> np.ndarray:
-        return self._trace[X].values[ix, :]
-
-    def get_fval_trace(self) -> Sequence[float]:
-        return self._trace[(FVAL, np.NaN)].values
-
-    def get_fval(self, ix) -> float:
-        return self._trace.loc[ix, (FVAL, np.NaN)]
-
-    def get_grad(self, ix) -> np.ndarray:
-        return self._trace[GRAD].values[ix, :]
-
-    def get_hess(self, ix) -> np.ndarray:
-        return self._trace[(HESS, np.NaN)].values[ix]
-
-    def get_res(self, ix) -> np.ndarray:
-        return self._trace[(RES, np.NaN)].values[ix]
-
-    def get_sres(self, ix) -> np.ndarray:
-        return self._trace[(SRES, np.NaN)].values[ix]
-
-    def get_chi2_trace(self) -> Sequence[np.ndarray]:
-        return self._trace[(CHI2, np.NaN)].values
-
-    def get_chi2(self, ix) -> float:
-        return self._trace.loc[ix, (CHI2, np.NaN)]
-
-    def get_schi2(self, ix) -> np.ndarray:
-        return self._trace[SCHI2].values[ix, :]
-
-    def get_time_trace(self) -> Sequence[float]:
-        return self._trace[(TIME, np.NaN)].values
-
-    def get_time(self, ix) -> float:
-        return self._trace.loc[ix, (TIME, np.NaN)]
+    @trace_wrap
+    def get_x_trace(
+            self, ix: Union[int, Sequence[int], None] = None
+    ) -> Union[Sequence[np.ndarray], np.ndarray]:
+        return self._trace[X].values[ix]
```

Comment:
This will return `np.ndarray` and not `Sequence[np.ndarray]` right? I am surprised that tests are not failing. I believe this previously led to failures in visualization routines.

Reply:
No tests failing. Do you remember which visualization routine? The different behavior of lists vs numpy arrays has led to problems in a few places already. Maybe it's safest to specify output as List here? Though it would be of interest what "List" vs "Sequence" specific behavior is used anywhere

Reply:
I think it was https://github.com/ICB-DCM/pyPESTO/blob/f74b4d0b54b2f159260664284da5e49709a8245a/pypesto/visualize/optimizer_history.py#L218

I would guess that `enumerate` is not equivalent for `np.ndarray` and `List`.

Reply:
Hm. Enumerate should only require an Iterable, which both implement -- in both cases should loop over the first dimension. Anyway, implementing as a list should be fine, efficiency issues should be unlikely.

Reply:
> This will return np.ndarray and not Sequence[np.ndarray] right? I am surprised that tests are not failing. I believe this previously led to failures in visualization routines.

Ok. It was meant to be a Sequence[np.ndarray], where the first dimension is just realized as a np.ndarray. I just realize that `issubclass(np.ndarray, collections.abc.Sequence)` is actually False. Don't really see the motivation, see also the thread here https://github.com/numpy/numpy/issues/2776.

Reply:
> I think it was
> 
> https://github.com/ICB-DCM/pyPESTO/blob/f74b4d0b54b2f159260664284da5e49709a8245a/pypesto/visualize/optimizer_history.py#L218
> 
> I would guess that `enumerate` is not equivalent for `np.ndarray` and `List`.

Just seeing: In the current implementation it should be ndarrays, not lists, too. https://github.com/ICB-DCM/pyPESTO/blob/develop/pypesto/objective/history.py#L719, so I guess that's fine?

Reply:
Hmm, those are all the getters for individual entries though, the getters for full traces should return lists of `np.ndarray`s though.